gpt4all-chat/src/database.cpp

```diff
@@ -1937,20 +2078,222 @@ QList<int> Database::searchEmbeddings(const std::vector<float> &query, const QLi
+            qWarning() << "Database ERROR: Failed to execute BM25 query:" << sqlQuery.lastError();
+            return {};
+        }
+
+        if (!sqlQuery.next())
+            continue;
+
+        bm25q = bm25Query;
+
+        do {
+            const int chunkId = sqlQuery.value(0).toInt();
+            const float score = sqlQuery.value(1).toFloat();
+            results.append({chunkId, score});
+        } while (sqlQuery.next());
+        break;
+    }
+
+    k = qMin(k, results.size());
+    std::partial_sort(
+        results.begin(), results.begin() + k, results.end(),
+        [](const SearchResult &a, const SearchResult &b) { return a.dist < b.dist; }
+    );
+
+    QList<int> chunkIds;
+    chunkIds.reserve(k);
+    for (int i = 0; i < k; i++)
+        chunkIds << results[i].chunkId;
+    return chunkIds;
+}
+
+float Database::computeBM25Weight(const Database::BM25Query &bm25q)
+{
+    float bmWeight = 0.0f;
+    if (bm25q.type == 0 /*exact*/) {
+        bmWeight = 0.9f; // the highest we give
+    } else {
+        // qlength is the length of the phrases in the query by number of distinct words
+        // ilength is the length of the natural language query by number of distinct words
+        // rlength is the number of stop words removed from the natural language query to form the query
+        float queryLengthWeight = powf(bm25q.qlength, 2) / powf(float(bm25q.ilength - bm25q.rlength), 2);
+        queryLengthWeight = qBound(0.0f, queryLengthWeight, 1.0f);
+        bmWeight = queryLengthWeight;
+        bmWeight = qBound(0.0f, bmWeight, 1.0f);
+        bmWeight = (0.25f + bmWeight * (0.75f - 0.25f));
+    }
+
+#if 0
+    qDebug()
+        << "bm25q.type"         << bm25q.type
+        << "bm25q.qlength"      << bm25q.qlength
+        << "bm25q.ilength"      << bm25q.ilength
+        << "bm25q.rlength"      << bm25q.rlength
+        << "bmWeight"           << bmWeight;
 #endif
 
-    std::vector<float> queryEmbd = m_embLLM->generateQueryEmbedding(text);
+    return bmWeight;
+}
+
+QList<int> Database::reciprocalRankFusion(const std::vector<float> &query, const QList<int> &embeddingResults,
+    const QList<int> &bm25Results, const BM25Query &bm25q, int k)
+{
+    // We default to the embedding results and augment with bm25 if any
+    QList<int> results = embeddingResults;
+
+    QList<int> missingScores;
+    QHash<int, int> bm25Ranks;
+    for (int i = 0; i < bm25Results.size(); ++i) {
+        if (!results.contains(bm25Results[i]))
+            missingScores.append(bm25Results[i]);
+        bm25Ranks[bm25Results[i]] = i + 1;
+    }
+
+    if (!missingScores.isEmpty()) {
+        QList<int> scored = scoreChunks(query, missingScores);
+        results << scored;
+    }
+
+    QHash<int, int> embeddingRanks;
+    for (int i = 0; i < results.size(); ++i)
+        embeddingRanks[results[i]] = i + 1;
+
+    const float bmWeight = bm25Results.isEmpty() ? 0 : computeBM25Weight(bm25q);
+    const int fusion_k = 60;
```

Comment:
Should cite the paper for the chosen value fusion_k and the algorithm in general.